### PR TITLE
Update toolchain.

### PR DESCRIPTION
**Pull request by egrimley-arm**
TODO:

- Update wasmtime and remove final reference to github.com/egrimley-arm

Draft log message for this PR:

Update Rust toolchain for Icecap and use stable compiler for rest.

Icecap now uses nightly-2021-08-25 and the rest uses stable 1.56.1.

In the new submodule icecap/sysroot/rust:

- Updates for internal API changes.
- A HACK: some locks are disabled.
- The submodule src/llvm-project is removed to save disc space.

In this repo changes include:

- Various features are removed to enable use of stable compiler.
- Crate versions updated, including: hex, lexical-core, x509-parser
- HACK to save disc space in icecap/Makefile: rm -rf ../sdk

### CODE REVIEW

**Comment by nspin** [comment]
Will this PR eventually turn into the one that we use to merge recent Rust support into Veracruz?

**Comment by egrimley-arm** [comment]
> Will this PR eventually turn into the one that we use to merge recent Rust support into Veracruz?

Yes, I hope. Lots of tidying to be done, and can we avoid checking out all of `rust` plus `llvm`?

**Comment by egrimley-arm** [comment]
https://github.com/veracruz-project/veracruz/runs/4533192689:
```
LLVM ERROR: IO failure on output stream: No space left on device
error: could not compile `serde`
warning: build failed, waiting for other jobs to finish...
error: couldn't create a temp dir: No space left on device (os error 28) at path "/tmp/nix-build-rustc-nightly.drv-0/source/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rmetaq91uih"
```

**Comment by nspin** [comment]
The issue here is that the Docker image that the CI for IceCap is running on is not primed with the right Nix store paths, so Nix is having to build a lot from scratch, ultimately exhausting disk space. The solution is to build and push a new IceCap CI Docker image, and refer to it in `.github/`.

See `veracruz/icecap/docker/Makefile` for details, or we can discuss 1-1.

**Comment by egrimley-arm** [comment]
`make ci-build` fails for me. That might be because I've had to point `icecap/icecap` at my own fork.

**Comment by dominic-mulligan-arm** [comment]
@egrimley-arm: what is the status of this PR?  Is it ready for review?

File: execution-engine/Cargo.toml
Comment (dominic-mulligan-arm) [inline execution-engine/Cargo.toml:35]: Can you put these under the `veracruz-project` rather than your personal Github account?

File: icecap/ci/run.sh
Comment (dominic-mulligan-arm) [inline icecap/ci/run.sh:1]: Can you put the standard authorship/licensing information as a comment at the top of this file?

File: icecap/sysroot/workspace/Cargo.toml
Comment (dominic-mulligan-arm) [inline icecap/sysroot/workspace/Cargo.toml:3]: Can you set the `version` number to `0.3.0`, set the `author` field to "The Veracruz development team", and add a short description in the `description` field?

File: policy-utils/Cargo.toml
Comment (dominic-mulligan-arm) [inline policy-utils/Cargo.toml:21]: Can this be moved under `veracruz-project`?

**Review by dominic-mulligan-arm — CHANGES_REQUESTED** [review]

File: execution-engine/Cargo.toml
Comment (egrimley-arm) [reply execution-engine/Cargo.toml:35]: I'm hoping https://github.com/veracruz-project/wasi-types/pull/6 can be merged first.

File: policy-utils/Cargo.toml
Comment (egrimley-arm) [reply policy-utils/Cargo.toml:21]: As above, I'm hoping https://github.com/veracruz-project/wasi-types/pull/6 can be merged first.

File: execution-engine/Cargo.toml
Comment (dominic-mulligan-arm) [reply execution-engine/Cargo.toml:35]: Ah OK, that makes sense.  I will take a look at that, now.

**Comment by dominic-mulligan-arm** [comment]
@egrimley-arm the PR removing the `proper` dependency from `wasi-types` has now been merged.

File: icecap/ci/run.sh
Comment (egrimley-arm) [reply icecap/ci/run.sh:1]: Done.

File: icecap/sysroot/workspace/Cargo.toml
Comment (egrimley-arm) [reply icecap/sysroot/workspace/Cargo.toml:3]: Done.

**Comment by egrimley-arm** [comment]
There is still one reference to `egrimley-arm`, for `wasmtime`. The required change in `wasmtime` is to delete four lines of the form `#![feature(...)]`. If I remember correctly, with those lines the code works with any "nightly" compiler (though you get a warning with more recent compilers), and without them the code works with any recent compiler, "nightly" or stable. Should we just make the change to `wasmtime` immediately before this PR is merged? (We already have `veracruz-v1` branch of `wasmtime`.)

**Comment by dominic-mulligan-arm** [comment]
Yes, I think we should merge the changes into our fork of `wasmtime`.  All forked dependencies need to be under the `veracruz-project` organization, rather than personal accounts.

Is there an existing PR for that?

**Comment by egrimley-arm** [comment]
> Yes, I think we should merge the changes into our fork of `wasmtime`. All forked dependencies need to be under the `veracruz-project` organization, rather than personal accounts.
> 
> Is there an existing PR for that?

https://github.com/veracruz-project/wasmtime/pull/9, which should be merged immediately before this PR
(updated to refer to `veracruz-project` instead of `egrimley-arm`).

**Review by dominic-mulligan-arm — COMMENTED** [review]
I think we're now in a position to proceed.  I've approved the `wasmtime` changes and modulo the change in the `wasmtime` dependency in this changeset, pointing it away from your repository and toward the project's, this is also now in a position where it is ready to be merged.

**Comment by egrimley-arm** [comment]
> I think we're now in a position to proceed. I've approved the `wasmtime` changes and modulo the change in the `wasmtime` dependency in this changeset, pointing it away from your repository and toward the project's, this is also now in a position where it is ready to be merged.

OK, let's do it then!

File: icecap/ci/run.sh
Comment (nspin) [inline icecap/ci/run.sh:28]: A fragment like this could be useful for debugging, but are we sure it belongs in `main`? I feel like it has an obfuscating effect. Perhaps we could get in the habit of temporarily inserting this into PRs when debugging them as necessary.

File: icecap/mk/realm.mk
Comment (nspin) [inline icecap/mk/realm.mk:47]: Fingerprints of these rlibs change, so repeated invocations of this target will result in an undesirable accumulation of stale rlibs. Which will rustc choose for building and linking against?

**Review by dominic-mulligan-arm — DISMISSED** [review]

File: icecap/Makefile
Comment (egrimley-arm) [reply icecap/Makefile]: Done.

File: icecap/Makefile
Comment (egrimley-arm) [reply icecap/Makefile:62]: I've done the makefile variable thing.

File: icecap/ci/run.sh
Comment (egrimley-arm) [reply icecap/ci/run.sh]: I've removed this fragment. It was already a separate PR anyway: https://github.com/veracruz-project/veracruz/pull/314

File: icecap/mk/realm.mk
Comment (egrimley-arm) [reply icecap/mk/realm.mk:47]: Let's fix this in a separate PR.

Comment (nspin) [reply icecap/mk/realm.mk:47]: 👍

File: icecap/Makefile
Comment (nspin) [reply icecap/Makefile:62]: Thanks

**Review by nspin — APPROVED** [review]

**Pull request merged**
